pkg/kindsys/customstructured.cue

```diff
@@ -0,0 +1,9 @@
+package kindsys
+
+// TODO
+#CustomStructured: {
+	#Structured
+	maturity: *"experimental" | "mature"
+	...
```

Comment:
This PR has done very little with `#CustomStructured`. i do expect it'll be _mostly_ same as `#CoreStructured`. That's borne out by the fact that most of the work in this PR has been in either `#Structured` or `_sharedKind` - and i actually did initially try to have them share a definition.

But we a number of known cases already where the constraints are different. [This example](https://github.com/grafana/grafana/pull/56492#issuecomment-1284791083) is just one.

Once i accepted that and leaned in, i found that designing for this had very nice benefits for the Go API - for example, funcs can clearly restrict their capabilities simply by [the set of metas they accept in their generic type parameter](https://github.com/grafana/grafana/pull/56492/files#diff-99bfecfb6b28f759f4e21b4931e140c1e3aa97a158b4fdcbab4c540181effc95R248).

Reply:
> It will have a fixed plugin/owner prefix? eg core can define "dashboard" but custom must be "k6-xyz"?

They will need namespacing at some level. That ties in with [the discussion](https://raintank-corp.slack.com/archives/C03BH67DS2J/p1666279859347609) @IfSentient and i were having today about how to handle "group" in k8s.

But i suspect we'll be able to express that directly in the schema when we get around to adding these as a field like this to the `pluginmeta` lineage:

```cue
  	// Unique name of the plugin. If the plugin is published on
	// grafana.com, then the plugin id has to follow the naming
	// conventions.
	id: string & strings.MinRunes(1)

	// ... other pluginmeta fields

	customTypes?: [string]: kindsys.#CustomStructured
```

Having `id` available means we can use it to establish a constraint relationship for all the `kindsys.#CustomStructured` values in `customTypes`. Same idea as what [this PR does with `machineName`](https://github.com/grafana/grafana/pull/56492/files#diff-fb814e775a368ffb53b5156eaa4d5a252647962b0017e3ab77bf4050526bc1d3R47)).

Maybe the constraints go on the custom type's `name` or maybe we make a new field. Doesn't really matter - as long as we're letting `plugin.cue` be an optional drop-in replacement for `plugin.json`, then the world is our oyster.

Reply:
> It looks like the difference here is around default maturity

The difference is both the default, and the allowable maturity stages:

* `#CoreStructured` allows `"committed" | "experimental" | "stable" | "mature"`, with `"committed"` as default
* `#CustomStructured` allows `"experimental" | "mature"`, with `"experimental"` as default

This is a function of trying to give [some clear meaning to maturity](https://docs.google.com/document/d/1gPzSCQn-0Qg86PYpPcR-l_kTQ4Ba1wgnNLZFKPSe8yQ/edit#heading=h.ajykcvt65mfb) - including public-facing guarantees - because only then can maturity milestones meaningfully reduce risk.

What's in this PR reflects what's in the doc - as [the table](https://docs.google.com/document/d/1gPzSCQn-0Qg86PYpPcR-l_kTQ4Ba1wgnNLZFKPSe8yQ/edit#heading=h.1bl5vdml7d4s) indicates, `committed` and `stable` are only currently planned to have meaning for core structured types. The only sane default i can think of is having unspecified be the first one for each kind category.